patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch

```diff
@@ -0,0 +1,695 @@
++  THR(::WaitForSingleObjectEx(
++          event_wait_for_media_capture_async_action_completion.Get(), INFINITE,
++          FALSE) == WAIT_OBJECT_0
++          ? S_OK
++          : E_FAIL);
++
++  // User granted access, good to go.
++  ReleaseInterface(ppMediaCapture);
++  *ppMediaCapture = media_capture.Detach();
++
++Cleanup:
++  return hr;
++}
++
+ uint32_t SafelyComputeMediaRatio(IMediaRatio* ratio_no_ref) {
+   HRESULT hr;
+   uint32_t numerator, denominator, media_ratio = 0;
+@@ -70,6 +226,31 @@ VideoType ToVideoType(const HString& sub_type) {
+   return converted_type;
+ }
+ 
++HRESULT GetMainViewDispatcher(ICoreDispatcher** main_view_dispatcher) {
++  HRESULT hr;
++
++  ComPtr<IActivationFactory> activation_factory;
++  ComPtr<ICoreApplication> coreApplication;
++  ComPtr<ICoreImmersiveApplication> immersiveApplication;
++  ComPtr<ICoreApplicationView> applicationView;
++  ComPtr<ICoreWindow> coreWindow;
++
++  THR(GetActivationFactory(
++      HStringReference(
++          RuntimeClass_Windows_ApplicationModel_Core_CoreApplication)
++          .Get(),
++      &activation_factory));
++
++  THR(activation_factory.As<ICoreApplication>(&coreApplication));
++  THR(coreApplication.As<ICoreImmersiveApplication>(&immersiveApplication));
++  THR(immersiveApplication->get_MainView(&applicationView));
++  THR(applicationView->get_CoreWindow(&coreWindow));
++  THR(coreWindow->get_Dispatcher(main_view_dispatcher));
++
++Cleanup:
++  return hr;
++}
++
+ HRESULT WaitForAsyncAction(ComPtr<IAsyncAction> async_action) {
+   HRESULT hr, hr_async_error;
+   const DWORD timeout_ms = 2000;
+diff --git a/modules/video_capture/windows/help_functions_winrt.h b/modules/video_capture/windows/help_functions_winrt.h
+index 0216d9b0ce..6b81bfc977 100644
+--- a/modules/video_capture/windows/help_functions_winrt.h
++++ b/modules/video_capture/windows/help_functions_winrt.h
+@@ -13,6 +13,7 @@
+ 
+ #include <windows.foundation.h>
+ #include <windows.media.mediaproperties.h>
++#include <windows.ui.core.h>
+ #include <wrl/client.h>
+ #include <wrl/event.h>
+ #include <wrl/wrappers/corewrappers.h>
+@@ -20,15 +21,51 @@
+ #include <cassert>
+ 
+ #include "modules/video_capture/video_capture_defines.h"
++#include "rtc_base/logging.h"
+ 
+ // Evaluates an expression returning a HRESULT.
+ // It jumps to a label named Cleanup on failure.
+-#define THR(expr)   \
+-  do {              \
+-    hr = (expr);    \
+-    if (FAILED(hr)) \
+-      goto Cleanup; \
++#ifdef _DEBUG
++#define THR(expr)                                           \
++  do {                                                      \
++    hr = (expr);                                            \
++    if (FAILED(hr)) {                                       \
++      RTC_LOG(LS_ERROR) << "HR: " << hr;                    \
++      RTC_LOG(LS_ERROR) << __FUNCTION__ << ":" << __LINE__; \
++      __debugbreak();                                       \
++      goto Cleanup;                                         \
```

Comment:
You'll find out that a guideline is just a guideline...
A quick search in Chromium code base found more than 3000+ uses of ``goto``. I didn't spend too much time looking at each use, but the beloved Protobuffer code has several uses of ``goto`` in its code and in the code that it generates. 

Including my favorite and so familiar:
``#define CHK_(x) if (PROTOBUF_PREDICT_FALSE(!(x))) goto failure``

Not all uses of ``goto`` are evil. I STRONGLY recommend a read of the chapter 16 of Steve McConnell's "Code Complete" book. Spoiler alert, page 351-352 will show you a very familiar code to this PR.

Reply:
@LoadLibrary the use of `goto`s and `THR` macros has already caused friction when contributing code to Google, see [here](https://webrtc-review.googlesource.com/c/src/+/167021) for example.

I'd say if we plan to contribute to a codebase, let's be respectful and follow the guidelines of that codebase as well as possible. As you point out, guidelines are just guidelines and this includes the COM error reporting conventions - we should use our growth mindset and evaluate constructs on their merits in a specific situation rather than stick to them because they have been successfully used elsewhere. IMHO contributing back this code is a concern, and we should make it as frictionless as possible, and complex macros and `goto`s are not _really_ essential here, should they be avoided.